### PR TITLE
sam - added recommendation entity, controller, and tests

### LINKED ISSUE #17: Add `DELETE` endpoint for a specific record in Recommendation table.
Add `DELETE` endpoint for a specific record in Recommendation table.

# Acceptance Criteria:

- [ ] In `RecommendationController.java` there is code for an 
      endpoint `DELETE /api/Recommendation?id=123` endpoint 
      that deletes the record if it exists, and returns 200 (ok) and 
      the text `record 123 deleted`, or returns 404 (Not Found) and
      the text `record 123 not found` if it does not.
- [ ] The Swagger-UI endpoints for this endpoint is well documented
      so that any member of the team can understand how to use it.
- [ ] The endpoint works as expected on localhost.
- [ ] The endpoint works as expected when deployed to Heroku.
- [ ] There is full test coverage (Jacoco) for the new code in 
      `RecommendationController.java`
- [ ] There is full mutation test coverage (Pitest) for new code in
      `RecommendationController.java`

### LINKED ISSUE #18: Add `PUT` (edit) endpoint for a single record in Recommendation table
Add `PUT` (edit) endpoint for a single record in Recommendation table

Note that it is a good idea to implement `GET` for a single item
first; when using `PUT` manually, it's convenient to be able to copy/paste the JSON representation of the data and then edit it in the swagger interface.

Eventually, it's typically the case that in order to populate an edit
form in the front end, you first do a `GET` on a particular item, 
and use that JSON to populate the fields in a form.  You can send the
JSON representation of the new item to the backend.

# Acceptance Criteria:

- [ ] In `RecommendationController.java` there is code for an 
      endpoint `PUT /api/Recommendation?id=123` endpoint 
      that accepts JSON for a new set of values for the database
      fields other than `id`, and updates the values of those fields.
- [ ] The Swagger-UI endpoints for this endpoint is well documented
      so that any member of the team can understand how to use it.
- [ ] The endpoint works as expected on localhost.
- [ ] The endpoint works as expected when deployed to Heroku.
- [ ] There is full test coverage (Jacoco) for the new code in 
      `RecommendationController.java`
- [ ] There is full mutation test coverage (Pitest) for new code in
      `RecommendationController.java`

### LINKED ISSUE #19: Add `GET` (show) endpoint for a single record in Recommendation table
Add `GET` (show) endpoint for a single record in Recommendation table

# Acceptance Criteria:

- [ ] In `RecommendationController.java` there is code for an 
      endpoint `GET /api/Recommendation?id=123` endpoint 
      that returns the JSON of the database record with id 123 if it
      exists, or a 400 and the error message `id 123 not found` if it
      does not.
- [ ] The Swagger-UI endpoints for this endpoint is well documented
      so that any member of the team can understand how to use it.
- [ ] The endpoint works as expected on localhost.
- [ ] The endpoint works as expected when deployed to Heroku.
- [ ] There is full test coverage (Jacoco) for the new code in 
      `RecommendationController.java`
- [ ] There is full mutation test coverage (Pitest) for new code in
      `RecommendationController.java`

### LINKED ISSUE #20: Create `RecommendationController`, add `GET` (index) and `POST` (create)
Create `RecommendationController`, add `GET` (index) and `POST` (create)

# Acceptance Criteria:

- [ ] There is a controller file `RecommendationController.java`
      in the expected directory.
- [ ] In `RecommendationController.java` there is 
      code for a `GET /api/Recommendation/all` endpoint 
      that returns a JSON list of all `Recommendation`s in the database.
      (We sometimes call this an *index* action since it lists all
      items in the database.)
- [ ] In `RecommendationController.java` there is 
      code for a `POST /api/Recommendation/post` endpoint
      that can be used to create a new entry in the table. (This
      is a *create* action.)
- [ ] The Swagger-UI endpoints for these are well documented so that
      any member of the team can understand what they are for and
      how to use them.
- [ ] The `POST` endpoint works as expected, in the sense that new
      records can be added to the database (on localhost).
- [ ] The `GET` endpoint works as expected, in the sense that the new
      records that are added show up (on localhost).
- [ ] The `GET` and `POST` endpoints work as expected when the 
      app is deployed to Heroku.
- [ ] There is full test coverage (Jacoco) for the methods in 
      `RecommendationController.java`
- [ ] There is full mutation test coverage (Pitest) for the methods in
      `RecommendationController.java`

### LINKED ISSUE #21: Create database table for Recommendation
Create database table for Recommendation

# Acceptance Criteria:

- [ ] There is an `@Entity` class called Recommendation.java
- [ ] There is a `@Repository` class called RecommendationRepository.java
- [ ] When you start up the repo on localhost, you can see the table
      using the H2 console (see the file `docs/h2-database.md` for 
      instructions.)
- [ ] When you deploy the app to Heroku, you can see the table
      using the psql command (See `Accessing Database Console` in the
      main `README.md` file for instructions.)

### PR DESCRIPTION
**Pull request by TokyoExpress**
Created new API endpoint with commands to get, set, update, and delete Recommendation Requests from the added database table.

### CODE REVIEW

File: src/test/java/edu/ucsb/cs156/example/controllers/RecommendationRequestsControllerTests.java [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+                LocalDateTime dateNeeded = LocalDateTime.parse("2022-01-05T00:00:00");
+
+
+                RecommendationRequest recommendationRequest = RecommendationRequest.builder()
+                                .requesterEmail("supman@gmail.com")
+                                .professorEmail("drbob@ucsb.edu")
+                                .explanation("idk")
+                                .dateRequested(dateRequested)
+                                .dateNeeded(dateNeeded)
+                                .done(false)
```
Comment (yuheho7749) [inline src/test/java/edu/ucsb/cs156/example/controllers/RecommendationRequestsControllerTests.java:99]: Need better tests

File: src/test/java/edu/ucsb/cs156/example/controllers/RecommendationRequestsControllerTests.java [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+                LocalDateTime dateRequested = LocalDateTime.parse("2022-01-03T00:00:00");
+                LocalDateTime dateNeeded = LocalDateTime.parse("2022-01-05T00:00:00");
+
+                RecommendationRequest recommendationRequest = RecommendationRequest.builder()
+                                .requesterEmail("supman@gmail.com")
+                                .professorEmail("drbob@ucsb.edu")
+                                .explanation("idk")
+                                .dateRequested(dateRequested)
+                                .dateNeeded(dateNeeded)
+                                .done(false)
```
Comment (yuheho7749) [inline src/test/java/edu/ucsb/cs156/example/controllers/RecommendationRequestsControllerTests.java:198]: Need better tests; fail mutation testing (do not rely on github actions; run mutation testing on local machine!)

File: src/test/java/edu/ucsb/cs156/example/controllers/RecommendationRequestsControllerTests.java [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+                                .build();
```
Comment (yuheho7749) [inline src/test/java/edu/ucsb/cs156/example/controllers/RecommendationRequestsControllerTests.java:284]: Need better tests (fails mutation testing)

File: src/test/java/edu/ucsb/cs156/example/controllers/RecommendationRequestsControllerTests.java [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+                RecommendationRequest recommendationRequestEdited = RecommendationRequest.builder()
+                                .requesterEmail("supman@gmail.com")
+                                .professorEmail("drbob@ucsb.edu")
+                                .explanation("idk just do it please")
+                                .dateRequested(dateRequested)
+                                .dateNeeded(dateNeeded)
+                                .done(true)
+                                .build();
```
Comment (yuheho7749) [inline src/test/java/edu/ucsb/cs156/example/controllers/RecommendationRequestsControllerTests.java]: Need better testing (fails mutation testing)

File: src/test/java/edu/ucsb/cs156/example/controllers/RecommendationRequestsControllerTests.java (hunk 2 above)
Comment (yuheho7749) [inline src/test/java/edu/ucsb/cs156/example/controllers/RecommendationRequestsControllerTests.java]: Need better tests; fail mutation testing (do not rely on github actions; run mutation testing on local machine!)

**Review by yuheho7749 — CHANGES_REQUESTED** [review]
Check mutation testing on local machine; do not rely on github actions to run the mutation testing

**Comment by TokyoExpress** [comment]
Fixed - mutation now passes locally.

**Review by yuheho7749 — APPROVED** [review]
LGTM
Sam, pls use/update the kanban board and link the corresponding issues that this PR will close (use the menu at the right under "Development" to link the issues)

**Review by jwangathan — APPROVED** [review]
LGTM all tests passed

**Pull request merged**